tools/pip-compile-wrapper.py

```diff
@@ -0,0 +1,300 @@
+        if package_name in package_names
+    ]
+    return {
+        key: {source_name for _, source_name in group}
+        for key, group in groupby(sorted(package_to_source, key=itemgetter(0)), key=itemgetter(0))
+    }
+
+
+def _get_requirement_package(source_name: str, target_package_name: str) -> Optional[str]:
+    for package_name, req_line in _get_requirement_packages(source_name):
+        if package_name == target_package_name:
+            return req_line
+    return None
+
+
+def _ensure_pip_tools() -> None:
+    if not which("pip-compile"):
+        print("pip-tools not available.")
+        pip_tools_req = _get_requirement_package(
+            REQUIREMENTS_SOURCE_DEV.replace(".in", ""), "pip-tools"
+        )
+        if not pip_tools_req:
+            raise RuntimeError(f"Package 'pip-tools' not found in {REQUIREMENTS_SOURCE_DEV}")
+        print(f"Installing {pip_tools_req}...", end="")
+        sys.stdout.flush()
+        process = subprocess.run(
+            [sys.executable, "-m", "pip", "install", pip_tools_req],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        process.check_returncode()
+        print("\b\b Done.")
+
+
+def compile_source(
+    upgrade_all: bool = False, verbose: bool = False, dry_run: bool = False
+) -> None:
+    for batch in _resolve_deps(SOURCE_PATHS.keys()):
+        for target in batch:
+            _run_pip_compile(target, upgrade_all=upgrade_all, verbose=verbose, dry_run=dry_run)
+
+
+def upgrade_source(
+    upgrade_package_names: Set[str], verbose: bool = False, dry_run: bool = False
+) -> None:
+    packages_to_sources = _get_sources_for_packages(upgrade_package_names, TargetType.ALL)
+
+    _newline = "\n  - "
+    missing_packages = upgrade_package_names - packages_to_sources.keys()
+    if missing_packages:
+        print(
+            "Some of the given packages were not found in either source or target files.\n"
+            "Please check that the packages are spelled correctly.\n"
+            "If any of these packages were newly added to any of the source files you need to "
+            f"run '{SCRIPT_NAME} compile' first.\n"
+            f"Missing package(s):\n  - {_newline.join(missing_packages)}"
+        )
+        sys.exit(1)
+    grouped_packages_to_sources = [
+        (set(package_name for package_name, _ in group), key)
+        for key, group in groupby(
+            sorted(packages_to_sources.items(), key=itemgetter(1)), key=itemgetter(1)
+        )
+    ]
+
+    for package_names, source_names in grouped_packages_to_sources:
+        print(f"Upgrading package(s):\n  - {_newline.join(package_names)}")
+        for batch in _resolve_deps(source_names):
+            for batch_source in batch:
+                _run_pip_compile(
+                    batch_source, upgrade_packages=package_names, verbose=verbose, dry_run=dry_run
+                )
+
+
+def main() -> None:
+    op = OptionParser(
```

Comment:
OptionParse is [deprecated](https://docs.python.org/3/library/optparse.html):

> Deprecated since version 3.2: The optparse module is deprecated and will not be developed further; development will continue with the argparse module.

Please use argparse, which also does what you need:

```python
parser = ArgumentParser()
parser.add_argument('-v', '--verbose', action='store_true')
parser.add_argument('-d', '--dry-run', action='store_true')
commands = parser.add_subparsers()
compile_parser = commands.add_parser('compile')
upgrade_parser = commands.add_parser('upgrade')
compile_parser.add_argument('files', nargs='*')
parser.parser()
```

Reply:
It would also be nice to have the target directory as an argument

Reply:
Right, totally forgot about argparse. 
Although it seems [optparse won't be going anywhere for a while](https://www.python.org/dev/peps/pep-0594/#optparse)

Reply:
> It would also be nice to have the target directory as an argument

Is is there a use case for that? 
There are some other places where it's assumed that source and targets files are in the same directory.

Reply:
> Is is there a use case for that?

It just remove some hardcode assumptions